### ISSUE TITLE
scrollIntoView is not a function

### ISSUE THREAD

**Issue by JakeDluhy** [tmpvar/jsdom#1695]
I'm running a test using a library (`grommet`) that uses `scrollIntoView` on an `a` tag. `jsdom` is throwing errors when hitting that line of code. I was able to get it to work by adding `window.HTMLElement.prototype.scrollIntoView = function() {};` in my testHelper file to setup the tests, but I thought I'd report it to see if it's something you want to fix.

**Comment by domenic** [comment]
Yeah this might be worth adding a not-implemented-warning stub for. We can't really implemented it since we don't do layout.

**Comment by QuantumInformation** [comment]
Any movement on this?

Trying to use in a Jest test, I'm trying to mount a component that calls scrollIntoView on an element, and the test always fails with

```
TypeError: node.scrollIntoView is not a function
        at Object.scrollInToView src/component/shared/foo.tsx:7
        at User.Object.<anonymous>.User.componentDidMount
```

**Comment by QuantumInformation** [comment]
^ @JakeDluhy how did you get HTMLElement to be defined in your jest test?

**Comment by enessoylu** [comment]
@QuantumInformation You can use this for the Jest:

`Element.prototype.scrollIntoView = jest.fn();`

**Comment by qmg-JamesT** [comment]
I see this fix everywhere, but were are you importing Element from?  I get 'Element is not defined' I am using enzyme with chai.

**Comment by QuantumInformation** [comment]
You need Element in your tsconfig libs:

ie
'dom'

**Comment by enessoylu** [comment]
@qmg-JamesT It is the global element class where the other elements are inherited **scrollIntoView** method from.

**Comment by ado-astpos** [comment]
Hi, is there any update on this without using jest?

**Comment by msmetana** [comment]
hello, you can use  **window.HTMLElement.prototype.scrollIntoView = jest.fn()**

**Comment by thisconnect** [comment]
I need the same for Element.prototype.scroll = jest.fn()

@enessoylu and @msmetana where do you recommend putting `Element.prototype.scrollIntoView = jest.fn()`, in each .test.

When using CRA there is a setupTest.ts, seems to work but idk

```
import '@testing-library/jest-dom/extend-expect'

Element.prototype.scroll = jest.fn()
```

**Comment by catalin-luntraru** [comment]
For `ava` or any other unit test runner you can use `window.HTMLElement.prototype.scrollIntoView = () => {}` and set it in `test.beforeEach()` or inside the test function

**Comment by tobiloeb** [comment]
For me it was enough to set:

``Element.prototype.scrollIntoView` = () => {};``

**once** into setupJest.ts

**Comment by jcampbell05** [comment]
This also seems to be the case with `scroll`

**Comment by ccloveak** [comment]
> @quantuminformation You can use this for the Jest:
> 
> `Element.prototype.scrollIntoView = jest.fn();`

Thanks! it worked.

**Comment by asharuDheen-code** [comment]
try this one..
window.HTMLElement.prototype.scrollIntoView = function() {};

**Comment by dobradovic** [comment]
How can I mock value from srcollIntoView? I'm using enzyme shallow / jest.

I'm getting TypeError: Cannot read property 'scrollIntoView' of null. 

Any advice would be helpful, I don't need to test anything regarding to scrollIntoView, I just need to mock some values, because my test failing and can not pass because scrollIntoView is null.

Thanks for your time!

**Comment by gabrielgaraujo** [comment]
Anyone have an idea why it happens?

**Comment by safi28** [comment]
> @quantuminformation You can use this for the Jest:
> 
> `Element.prototype.scrollIntoView = jest.fn();`

Thank you it works!

**Comment by ledenis** [comment]
Why not add the not-implemented-warning stub as suggested in the second comment?

Is there a reason why it has been done for `scrollTo` but not `scrollIntoView`? https://github.com/jsdom/jsdom/blob/65ef06f3771a33dcbf7bef442fdc7e07816d008d/lib/jsdom/browser/Window.js#L931

**Comment by trajano** [comment]
FWIW it happy-dom appears to handle this.

**Comment by rgant** [comment]
It would be quite nice if #3639 could be merged so that I can replace `scrollIntoView` with a spy instead of having to manually add and remove it from the Element prototype.

**Comment by agrberg** [comment]
For anyone landing here from a search, sharing a recent failure mode plus the project-wide workaround we landed on.

We had a controller method that called `scrollIntoView` from inside an `async` function wrapped in `try/catch`. In jsdom that call throws, but because the throw was caught by our own error handler the exception never surfaced. What we saw instead was a downstream symptom. A deduplication guard that depended on state set just after the `scrollIntoView` call had not run, so a function we expected to be invoked once was being invoked twice. It took longer than it should have to connect those dots back to `scrollIntoView`, which is the real failure mode worth flagging. If anyone here is debugging a "mock called more times than expected" mystery in jsdom, this is worth ruling out.

For the fix, rather than stubbing in every test's `beforeEach` (easy to forget, and forgetting reintroduces exactly the silent-catch trap above), we moved the stub into a single Jest setup file referenced from `setupFilesAfterEnv`:

```js
// jest.setup.js
Object.defineProperty(window.HTMLElement.prototype, 'scrollIntoView', {
  value: jest.fn(),
  writable: true,
  configurable: true,
});
```

```js
// jest.config.js
module.exports = {
  testEnvironment: 'jest-environment-jsdom',
  setupFilesAfterEnv: ['<rootDir>/jest.setup.js'],
};
```

Two small notes that tripped us up. `setupFilesAfterEnv` is the right hook rather than `setupFiles`, because `jest.fn` is not available in the latter (the framework has not loaded yet). And using `defineProperty` with `configurable: true` means individual tests can still call `jest.spyOn(HTMLElement.prototype, 'scrollIntoView')` when they need to assert on calls. The same setup file is a good place for the other "missing in jsdom" stubs that tend to cluster, like `IntersectionObserver`, `ResizeObserver`, and `matchMedia`.

A separate thought for the maintainers. Even just marking the method as "not implemented" with a clearer warning (the spirit of #3639) would have made our silent-catch case noisy in the console rather than invisible, which is the bug-finding behavior most testers actually want, and would not commit jsdom to any layout work.
